packages/bp/src/core/dialog/dialog-engine.ts

```diff
@@ -219,79 +268,88 @@ export class DialogEngine {
-      try {
-        return this._findNode(botId, flow, nodeName)
-      } catch (err) {
-        // ignore
-      }
-      return undefined
-    }
-    const findFlowWithoutError = flowName => {
-      try {
-        return this._findFlow(botId, flowName)
-      } catch (err) {
-        // ignore
-      }
-      return undefined
-    }
-
     const currentFlow = this._findFlow(botId, event.state.context?.currentFlow!)
-    const currentNode = findNodeWithoutError(currentFlow, event.state.context?.currentNode)
+    const currentNode = this.findNodeWithoutError(botId, currentFlow, event.state.context?.currentNode)
 
     // Check for a timeout property in the current node
     let timeoutNode = _.get(currentNode, 'timeout')
     let timeoutFlow: FlowView | undefined = currentFlow
 
     // Check for a timeout node in the current flow
     if (!timeoutNode) {
-      timeoutNode = findNodeWithoutError(currentFlow, 'timeout')
+      timeoutNode = this.findNodeWithoutError(botId, currentFlow, 'timeout')
     }
 
     // Check for a timeout property in the current flow
     if (!timeoutNode) {
       const timeoutNodeName = _.get(timeoutFlow, 'timeoutNode')
       if (timeoutNodeName) {
-        timeoutNode = findNodeWithoutError(timeoutFlow, timeoutNodeName)
+        timeoutNode = this.findNodeWithoutError(botId, timeoutFlow, timeoutNodeName)
       }
     }
 
     // Check for a timeout.flow.json and get the start node
     if (!timeoutNode) {
-      timeoutFlow = findFlowWithoutError('timeout.flow.json')
+      timeoutFlow = this.findFlowWithoutError(botId, 'timeout.flow.json')
       if (timeoutFlow) {
         const startNodeName = timeoutFlow.startNode
-        timeoutNode = findNodeWithoutError(timeoutFlow, startNodeName)
+        timeoutNode = this.findNodeWithoutError(botId, timeoutFlow, startNodeName)
       }
     }
 
     if (!timeoutNode || !timeoutFlow) {
       throw new TimeoutNodeNotFound(`Could not find any timeout node or flow for session "${sessionId}"`)
     }
 
-    if (event.state.context) {
-      event.state.context.currentNode = timeoutNode.name
-      event.state.context.currentFlow = timeoutFlow.name
-      event.state.context.queue = undefined
-      event.state.context.previousFlow = currentFlow.name
-      event.state.context.previousNode = currentNode?.name as string
-      event.state.context.jumpPoints = [
-        ...(event.state.context?.jumpPoints || []),
-        {
-          flow: currentFlow.name,
-          node: currentNode?.name as string
-        }
-      ]
-      event.state.context.hasJumped = true
-    }
+    this.fillContextForTransition(event, { currentFlow, currentNode, nextFlow: timeoutFlow, nextNode: timeoutNode })
 
     return this.processEvent(sessionId, event)
   }
 
-  private _endFlow(event: IO.IncomingEvent) {
-    event.state.context = {}
-    event.state.temp = {}
+  public async processConversationEnd(botId: string, sessionId: string, event: IO.IncomingEvent) {
```

Comment:
Can you add typings to the return value ?

Reply:
changed